app/Services/LoanService.php

```diff
@@ -32,10 +75,64 @@ public function createLoan(User $user, int $amount, string $currencyCode, int $t
      * @param  string  $currencyCode
      * @param  string  $receivedAt
      *
-     * @return ReceivedRepayment
+     * @return ReceivedRepayment|Model
      */
     public function repayLoan(Loan $loan, int $amount, string $currencyCode, string $receivedAt): ReceivedRepayment
     {
-        //
+        $receivedRepayment = $loan->receivedRepayments()->create([
+            'amount' => $amount,
+            'currency_code' => $currencyCode,
+            'received_at' => Carbon::createFromFormat('Y-m-d', $receivedAt)->format('Y-m-d'),
+        ]);
+        if ($receivedRepayment) {
+            Log::info('Store received repayment success!');
+            Log::debug('Received repayment data: ' . json_encode($receivedRepayment));
+
+            $scheduleRepayments = $loan->scheduledRepayments()
+                ->where('status', ScheduledRepayment::STATUS_DUE)
+                ->orderBy('due_date')->get();
+
+            self::recursiveScheduleRepaymentUpdate($scheduleRepayments, $amount);
+            $loan->outstanding_amount = $loan->scheduledRepayments()
+                ->where('status', '<>', Loan::STATUS_REPAID)->sum('outstanding_amount');
+            if ($loan->outstanding_amount == 0) {
+                $loan->status = Loan::STATUS_REPAID;
+            }
+            $loan->save();
+        } else {
+            Log::error('Store received repayment was failed!');
```

Comment:
this should be an exception

Reply:
yes, I was in too much of a hurry to can put it in a transaction